userspace/libsinsp/sinsp_events.h

```diff
@@ -0,0 +1,369 @@
+#pragma once
+
+#include "event.h"
+#include <utils.h>
+#include <sinsp_exception.h>
+#include <sinsp_public.h>
+#include <unordered_set>
+#include <string>
+#include <vector>
+#include <functional>
+
+namespace libsinsp {
+namespace events {
+
+template<typename ppm_type>
+class set
+{
+private:
+	using vec_t = std::vector<uint8_t>;
+	vec_t m_types{};
+	ppm_type max;
+
+	inline void check_range(ppm_type e) const
+	{
+		if(e > max)
+		{
+			throw sinsp_exception("invalid event type");
+		}
+	}
+
+public:
+	set(set&&)  noexcept = default;
+	set(const set&) = default;
+	set& operator=(set&&)  noexcept = default;
+	set& operator=(const set&) = default;
+	set<ppm_type>() = delete;
+
+	inline explicit set(ppm_type maxLen):
+		m_types(maxLen + 1, 0),
+		max(maxLen)
+	{
+	}
+
+	inline explicit set(std::unordered_set<ppm_type> vals)
+	{
+		set<ppm_type>();
+		for (const auto &val : vals)
+		{
+			insert(val);
+		}
+	}
+
+	inline void insert(ppm_type e)
+	{
+		check_range(e);
+		m_types[e] = 1;
+	}
+
+	inline void remove(ppm_type e)
+	{
+		check_range(e);
+		m_types[e] = 0;
+	}
+
+	void merge(const set& other)
+	{
+		if (other.max != max)
+		{
+			throw sinsp_exception("cannot merge different set");
+		}
+		for(size_t i = 0; i <= max; ++i)
+		{
+			m_types[i] |= other.m_types[i];
+		}
+	}
+
+	void merge(const std::set<ppm_type>& other)
+	{
+		if (other.size() != max)
+		{
+			throw sinsp_exception("cannot merge different set");
+		}
+		for(const auto& e : other)
+		{
+			insert(e);
+		}
+	}
+
+	inline bool contains(ppm_type e) const
+	{
+		check_range(e);
+		return m_types[e] != 0;
+	}
+
+	void clear()
+	{
+		for(auto& v : m_types)
+		{
+			v = 0;
+		}
+	}
+
+	bool empty() const
+	{
+		return equals(set<ppm_type>{});
+	}
+
+	bool equals(const set& other) const
+	{
+		return m_types == other.m_types;
+	}
+
+	set diff(const set& other)
+	{
+		if (other.max != max)
+		{
+			throw sinsp_exception("cannot diff different set");
+		}
+		set<ppm_type> ret(max);
+		for(size_t i = 0; i <= max; ++i)
+		{
+			ret.m_types[i] = m_types[i] ^ other.m_types[i];
+		}
+		return ret;
+	}
+
+	set intersect(const set& other)
+	{
```

Comment:
I know we have `merge` that does a similar thing, but at this point would it make sense to add `set union(const set& other) const`? That way this class would be all-around versatile.

Reply:
Yep fully agree!